rust/relay/ebpf-turn-router/src/move_payload.rs

```diff
@@ -0,0 +1,248 @@
+        return Err(Error::PacketTooShort);
+    }
+
+    // Extend tail
+    let ret = unsafe { bpf_xdp_adjust_tail(ctx.ctx, CdHdr::LEN as i32) };
+    if ret < 0 {
+        return Err(Error::XdpAdjustTailFailed(ret));
+    }
+
+    let data_size = ctx.data_end() - ctx.data();
+
+    // Copy payload
+    let original_data_size = data_size - CdHdr::LEN;
+    let payload_size = original_data_size - off;
+    let src_off = off as i32;
+    let dst_off = (off + CdHdr::LEN) as i32;
+
+    copy_payload(ctx, src_off, dst_off, payload_size)?;
+
+    // Write header
+    let data_start = ctx.data();
+    let data_end = ctx.data_end();
+
+    // Verify the header write location is within bounds
+    if data_start + off + CdHdr::LEN > data_end {
+        return Err(Error::PacketTooShort);
+    }
+
+    let src_ptr = header as *const CdHdr;
+    let dst_ptr = data_start + off;
+    unsafe {
+        core::ptr::copy_nonoverlapping(src_ptr as *const u8, dst_ptr as *mut u8, CdHdr::LEN);
+    }
+
+    Ok(())
+}
+
+/// Slide payload toward head by `CdHdr::LEN` and shrink tail by `CdHdr::LEN`.
+#[inline(never)]
+fn shift_forward_and_shrink<const IP_HEADER_LEN: usize>(ctx: &XdpContext) -> Result<(), Error> {
+    let off = payload_offset::<IP_HEADER_LEN>();
+
+    if ctx.data() + off > ctx.data_end() {
+        return Err(Error::PacketTooShort);
+    }
+
+    let data_size = ctx.data_end() - ctx.data();
+
+    // Copy payload
+    let remaining = data_size - off - CdHdr::LEN;
+    let src_off = (off + CdHdr::LEN) as i32;
+    let dst_off = off as i32;
+
+    copy_payload(ctx, src_off, dst_off, remaining)?;
+
+    // Trim the now-unused bytes from the tail
+    let ret = unsafe { bpf_xdp_adjust_tail(ctx.ctx, -(CdHdr::LEN as i32)) };
+    if ret < 0 {
+        return Err(Error::XdpAdjustTailFailed(ret));
+    }
+
+    Ok(())
+}
+
+#[inline(always)]
+fn payload_offset<const IP_HEADER_LEN: usize>() -> usize {
+    EthHdr::LEN + IP_HEADER_LEN + UdpHdr::LEN
+}
+
+/// Copy payload by `n` bytes with direction determined by src/dst relationship.
+/// When dst > src, we copy forward (ascending) to avoid overlaps.
+/// When dst < src, we copy backward (descending) to avoid overlaps.
+///
+/// This is performed in a single pass using binary decomposition from 1024 down to 1 byte.
+/// This done because the verifier doesn't like non compile-time guarantees that pointers
+/// are within bounds, and the variable length payload can cause an issue.
+#[inline(never)]
+fn copy_payload(
+    ctx: &XdpContext,
+    mut src_off: i32,
+    mut dst_off: i32,
+    n: usize,
+) -> Result<(), Error> {
+    if n > MAX_PAYLOAD {
+        return Err(Error::PacketTooLong);
+    }
+
+    // Get data pointers once at the start
+    let data_start = ctx.data() as *const u8;
+    let data_end = ctx.data_end() as *const u8;
+
+    // Determine direction: -1 for backward (when dst > src), 1 for forward
+    let dir: i32 = if dst_off > src_off { -1 } else { 1 };
```

Comment:
Can this decision not be part of `copy_n`? Then you wouldn't have to repeat the arithemtics for each step.

Reply:
I'll try push a commit later to show what I mean.